sdks/python/apache_beam/examples/inference/README.md

```diff
@@ -154,6 +169,52 @@ This writes the output to the `predictions.csv` with contents like:
 ```
 Each line has data separated by a semicolon ";". The first item is the file name. The second item is a list of predicted instances.
 
+---
+## Object Detection
+
+[`tensorrt_object_detection.py`](./tensorrt_object_detection.py) contains an implementation for a RunInference pipeline that performs object detection using [Tensorflow Object Detection's](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md) SSD MobileNet v2 320x320 architecture.
+
+The pipeline reads the images, performs basic preprocessing, passes them to the TensorRT implementation of RunInference, and then writes the predictions to a text file.
+
+### Dataset and model for image classification
+
+You will need to create or download images, and place them into your `IMAGES_DIR` directory. Popular dataset for such task is [COCO dataset](https://cocodataset.org/#home). COCO validation dataset can be obtained [here](http://images.cocodataset.org/zips/val2017.zip).
+- **Required**: A path to a file called `IMAGE_FILE_NAMES` that contains the absolute paths of each of the images in `IMAGES_DIR` on which you want to run image segmentation. Paths can be different types of URIs such as your local file system, a AWS S3 bucket or GCP Cloud Storage bucket. For example:
+```
+/absolute/path/to/000000000139.jpg
+/absolute/path/to/000000289594.jpg
+```
+- **Required**: A path to a file called `TRT_ENGINE` that contains the pre-built TensorRT engine from SSD MobileNet v2 320x320 model. You will need to [follow instructions](https://github.com/NVIDIA/TensorRT/tree/main/samples/python/tensorflow_object_detection_api) on how to download and convert this SSD model into TensorRT engine. At [Create ONNX Graph](https://github.com/NVIDIA/TensorRT/tree/main/samples/python/tensorflow_object_detection_api#create-onnx-graph) step, keep batch size at 1. As soon as you are done with [Build TensorRT Engine](https://github.com/NVIDIA/TensorRT/tree/main/samples/python/tensorflow_object_detection_api#build-tensorrt-engine) step. You can use resulted engine as `TRT_ENGINE` input. In addition, make sure that environment you use for TensorRT engine creation is the same environment you use to run TensorRT inference. It is related not only to TensorRT version, but also to a specific GPU used. Read more about it [here](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#compatibility-serialized-engines).
+
+- **Required**: A path to a file called `OUTPUT`, to which the pipeline will write the predictions.
+- **Optional**: `IMAGES_DIR`, which is the path to the directory where images are stored. Not required if image names in the input file `IMAGE_FILE_NAMES` have absolute paths.
+
+### Running `tensorrt_object_detection.py`
+
+To run the image classification pipeline locally, use the following command:
```

Comment:
1. I assume one will need a NVIDIA GPU to successfully run this example.  I think we should mention it? Or without GPUs one can still run the example, but with a degraded performance? 

2. Related to 1: if we want to run the pipeline on the dataflow runner: will users still need a machine with GPU at _job submission time_? For example, to be able to install all the tensorrt dependencies and submit the pipeline for execution on Dataflow?

Reply:
1. README.md has a section about TRT dependencies. As soon as users will go through it, they will realize that they need a certain set of requirements. Not only software but hardware. In addition, it says: "TensorRT SDK for high-performance deep learning inference with NVIDIA GPUs". I think it is clear enough. Please let me know if you think otherwise and would like to change something. TensorRT only works with GPUs, there is no other version. 

2. You can install driver, CUDA, cuDNN, TRT without GPU being present. However, if you want to run something with TRT. You will need a GPU. Additional [good info](https://github.com/apache/beam/pull/22131#issuecomment-1189323009), cannot install trt or cuda python packages without CUDA library and TensorRT library.

Reply:
Re#2: 

If at all possible we should import the libraries requiring GPUs only in the portion of the code that will be executed on the worker. It would be a much more convenient user experience if a user does not need GPUs to launch a pipeline that will spin up workers with GPUs in the cloud.

Reply:
This might mean considering moving some imports within functions, instead of having them at the top of the file.

Reply:
What is executed on the workers and what is executed on a submission machine? I am not very familiar since I pretty much just ran unit tests and obj detection sample locally. Considering we can do this, how would we do smth like this? Can you give an example? Maybe Pytorch implementation handles it somehow, I don't think you need any pytroch packages (or any packages) on a submission machine. 

@yeandy You were testing on GCP, any interesting observations/comments about it? Maybe something you can recommend? (Btw, congrats on presenting at the [summit](https://2022.beamsummit.org/sessions/runinference/), saw photos, looked cool)

Reply:
Ok, you partially answered my question. It will make a mess though, every function in `tensorrt_inference.py` needs TRT.  Plus we need to maintain global trt logger. Maybe we can package it somehow and bounce it around.

Reply:
Roughly speaking, at submission we create a DAG of the pipeline. The transforms and pipeline user code is represented in serialized form. 
On the runner the code will be deserialized and actually executed run.

To test out the UX, one should try to run the example on Dataflow runner. First, launching from a machine with GPUs, then launch from a machine without GPUs. I assume we have already ran with the Direct  runner (all processing done on the same machine), if not do that first.

Reply:
Maybe you can propose a solution that you would like to see. I'll verify it works on my end and we can commit it.

Reply:
Re summit, thanks @azhurkevich!

---

Per my GCP testing, nothing to call out. Though I've only had a chance to run the unit tests, which didn't require Dataflow. FYI, this is what I did to run the unit tests:
- Create GCE VM: Choose 2 `NVIDIA Tesla T4` GPUs and also `n1-standard-2` (2 vCPU, 7.5 GB memory) for machine type. Choose “Debian 10 based Deep Learning VM with CUDA 11.3 M94” for Boot Disk.
- Build docker inside VM: `docker build -f Dev_Beam.dockerfile -t dev_beam .`
- Clone Alexander's apache_beam fork and checkout to the `tensorrt_runinference` branch.
- Build tensorrt engine and upload those files (and onnx files) to the `apache-beam-ml` GCS bucket.
- Run tests: `docker run --rm -it --gpus all -v /home/yeandy/:/mnt -w /mnt/beam/sdks/python dev_beam python setup.py test -s apache_beam.ml.inference.tensorrt_inference_test`

I'm catching up with @AnandInguva tomorrow to clarify our infra setup for these tests.

---

Also caught up with @tvalentyn. I'm going to work on running the `tensorrt_object_detection` example on Dataflow by launching from both a machine with GPUs and a machine without GPUs, and see what code refactoring is necessary. This will require a bit of tinkering, which may take a few days. I will give an update as soon as I can.

Reply:
Update:

I'm able to kick off a DataFlow job for the `tensorrt_object_detection` pipeline from my local machine. It runs successfully, reading an input txt file and trt engine file from GCS, as well as writing predictions to GCS, with no issue. 

There may be cleaner ways to do this, so please let me know if you have any ideas. Here are the changes.

1. Move any `tensorrt` or `cuda` imports inside helper functions
2. Wrap `TRT_LOGGER = trt.Logger(trt.Logger.INFO)` inside a try/catch block so that I could get the job submission to work.
3. Move `trt.init_libnvinfer_plugins(TRT_LOGGER, namespace="")` from the `__init__` function of `TensorRTEngineHandlerNumPy` to the `load_model` part. `trt.init_libnvinfer_plugins(TRT_LOGGER, namespace="")` needs to be called once, but we can't leave it in `__init__` because logic in here will be called at job submission time, which won't guarantee to have `tensorrt` installed locally.
4. Add `from __future__ import annotations` to allow type-checking to pass. Otherwise, we will see `NameError: name 'trt' is not defined` in this line: `def __init__(self, engine: trt.ICudaEngine):` 
5. Slightly modified the Dockerfile to reintroduce `ENTRYPOINT [ "/opt/apache/beam/boot" ]` so that the pipeline could start.

In addition. One thing we need to do is to be able to rebuild the TRT engine from the onnx file (already in GCS) every time. For this job, I did the engine building in a separate VM. The next thing to do is to create a job -- probably a Dataflow one -- to build the engine before running the actual test. 

Please see this [PR](https://github.com/azhurkevich/beam/pull/1) with my changes, and let me know what you think.

Reply:
@yeandy Thanks for this crystal clear explanation! In the end, I just want to make sure that whatever I do is compliant with your vision. From the high level looks good to me, [I added a small suggestion](https://github.com/azhurkevich/beam/pull/1#issuecomment-1201944306). We can merge after it is addressed. 
I have diving deeper and figuring out GCP planned for this week. I am complete rookie so if you make any Dataflow suggestions, please be specific so that I can understand.

Reply:
Updated the PR per your feedback. Tested, and it still works. Thanks!

And for dataflow, first make sure you have a GCP account, and can access your console and navigate to the Dataflow tab. I would suggest following [this](https://cloud.google.com/dataflow/docs/quickstarts/create-pipeline-python) or [this](https://beam.apache.org/get-started/quickstart-py/) (they are pretty similar) quickstart guide.

Reply:
@yeandy Merged